Blockmanager ui

squito opened the pull request:
This is an update to https://github.com/mesos/spark/pull/295

I didn't add any additional functionality beyond just bringing it up to date w/ master; I agree additional changes would be nice, but wanted to get this much merged in at least.

Comments greatly appreciated -- some parts of the merge were a bit ugly, and this section of the code is totally new to me.

dennybritz committed:
BlockManager UI.

dennybritz committed:
code Formatting

dennybritz committed:
Nicer storage level format on RDD page

dennybritz committed:
Merge branch 'master' into blockmanagerUI

Conflicts:
	core/src/main/twirl/spark/deploy/worker/index.scala.html

dennybritz committed:
Merge branch 'master' into blockmanagerUI

dennybritz committed:
Refactor BlockManager UI and adding worker details.

dennybritz committed:
Merge branch 'master' into blockmanagerUI

Commit:
Merge remote-tracking branch 'dennybritz/blockmanagerUI' into blockmanager_ui

Conflicts:
	core/src/main/scala/spark/RDD.scala
	core/src/main/scala/spark/storage/BlockManagerMaster.scala
	core/src/main/scala/spark/storage/StorageLevel.scala

Commit:
Fix up some problems from the merge

Commit:
Merge branch 'master' into blockmanager_ui

Conflicts:
	core/src/main/scala/spark/RDD.scala

pwendell commented on core/src/main/scala/spark/RDD.scala:
```diff
@@ -94,6 +94,9 @@ abstract class RDD[T: ClassManifest](
   /** How this RDD depends on any parent RDDs. */
   protected def getDependencies(): List[Dependency[_]] = dependencies_
 
+  // A friendly name for this RDD
```
Nit: use /*\* */ style comments as in the rest of this file

pwendell commented on core/src/main/scala/spark/RDD.scala:
```diff
@@ -108,7 +111,13 @@ abstract class RDD[T: ClassManifest](
   val id = sc.newRddId()
 
-  /**
+  /* Assign a name to this RDD */
```
Nit: use /*\* */ style comments as in the rest of this file

pwendell commented on core/src/main/scala/spark/RDD.scala:
```diff
@@ -119,6 +128,8 @@ abstract class RDD[T: ClassManifest](
     }
     storageLevel = newLevel
+    // Register the RDD with the SparkContext
+    sc.persistentRdds(id) = this
```
One issue with the introduction of streaming is how to deal with state that accumulates over the course of a long-running job. In this case, it seems like this particular HashMap only gets larger, which will pose a problem. I'm going to defer to TD or Matei here to figure out how to best address this particular case. TD has a handful of MetaDataCleaner implementations which go through and delete things based on a timestamp (checkout DAGScheduler), that may be appropriate to use here.

pwendell replied:
Just looked closer, and the best way is definitely to use the special HashMap and cleaner that are used in DAGScheduler.

JoshRosen replied:
Is simply purging values that are older than a threshold the right approach here?  It seems like most of the space overhead would come from the RDDs rather than the integer HashMap keys.  Purging old values will only benefit us if we can garbage-collect those values. 

Imagine that I have a job that runs hundreds of iterations of an algorithm over a cached data set, which might take a long time to complete.  A naive timeout would purge this cached long-lived RDD from the HashMap even though there are other references keeping it alive, so we unnecessarily lose the ability to display its information.

On the other hand, we don't want to add strong references that would prevent RDDs from being garbage-collected.

Maybe we could use a WeakConcurrentHashMap, which would address this problem.  It looks like Guava's [MapMaker](http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/collect/MapMaker.html) can construct these, but I don't have personal experience with it.

This approach doesn't address cases where the cache contains blocks from an RDD that has been garbage-collected.  I suppose that we could use MapMaker to make a map that uses soft references, but I don't know how this performs in practice.

pwendell replied:
How to deal with cleaning meta-data correctly is a complicated issue and one which is not yet completely addressed in the codebase. The reason this is necessary is streaming jobs which might create thousands of RDD's over a short period of time, so any state really needs to be garbage collected periodically. Otherwise hashmaps like this bloat to the point where they start to impact performance.

For now, TD has introduced these helper data structures which use a timeout based approach. He's added them throughout the codebase, but they are enabled only in streaming mode. In normal batch mode, it acts just like any normal concurrent hashmap. 

Figuring out the correct GC behavior for all types of state is something TD is working on right now, but in the meantime this timeout approach is being employed.

pwendell commented on core/src/main/scala/spark/storage/BlockManagerUI.scala:
```diff
@@ -0,0 +1,86 @@
+
+        val rdds = StorageUtils.rddInfoFromStorageStatus(storageStatusList, sc)
+
+        spark.storage.html.index.render(maxMem, remainingMem, diskSpaceUsed, rdds, storageStatusList)
```
This line exceeds 100 characters.

pwendell commented on core/src/main/scala/spark/storage/BlockManagerUI.scala:
```diff
@@ -0,0 +1,86 @@
+
+
+        val storageStatusList = status.asInstanceOf[ArrayBuffer[StorageStatus]].toArray
+        val filteredStorageStatusList = StorageUtils.filterStorageStatusByPrefix(storageStatusList, prefix)
```
> 100 characters

pwendell commented:
Thanks for bringing this up to date! Overall, looks good, with a few small nits. Would be great to get this in as a base for future UI work.

Commit:
code reformatting

Commit:
switch to TimeStampedHashMap for storing persistent Rdds

squito commented:
I fixed the formatting issues, and changed persistentRdds to be a TimeStampedHashMap.  However, its not getting cleaned up anywhere yet -- I wasn't sure if I should move that logic up from DAGScheduler into SparkContext, push the map down to DAGScheduler, or some other refactoring ... thoughts?

Also, as I've played around with this a bit, it seems odd to me that the UI is hosted by the driver, not the master.  (At the very least, the master UI should link to the Storage UI.)  maybe its ok as is for the moment, though ...

squito commented:
oops, sorry this is not ready yet, some stupid mistakes ...

Commit:
fixup  1cadaa1, changed api of map

squito commented:
fixed the compile errors, but still not sure about the where to clear the persistentRdd values

Commit:
add metadatacleaner for persisentRdd map

squito commented:
I think this is ready to go now.

(The maven build worked, but I had to give it more memory "MAVEN_OPTS=-Xmx1024m mvn -e compile", not sure if that has always been the case, or something from this change ...)

pwendell commented:
Looks good to me! Thanks for the contribution.

mateiz commented:
Alright, merged it. Thanks.

stephenh commented on core/src/main/scala/spark/SparkContext.scala:
```diff
@@ -645,6 +658,12 @@ class SparkContext(
+  private[spark] def cleanup(cleanupTime: Long) {
+    var sizeBefore = persistentRdds.size
+    persistentRdds.clearOldValues(cleanupTime)
+    logInfo("idToStage " + sizeBefore + " --> " + persistentRdds.size)
```
idToStage? Copy/paste error?

mateiz replied:
Hmm yeah good catch.. I'll fix that
